### DISCUSSION TITLE
How to do disaster recovery?

### DISCUSSION THREAD

**Discussion by Gomez12** [ravendb/ravendb#16073] [Q&A] [answered]
We are looking at a worst-case disaster recovery scenario for our existing ravendb database.
We especially like the functionality that you can have many map-reduces for different reports etc. Which is al pre-calculated then.

But now it comes to bite us in the ass it looks like. As it handles the daily load perfectly.
But if we try a disaster recovery where we would for example need to re-index the complete data then it completely overloads our storage I/O capacity in such a way that it can hardly go through the indexing job.
Everything comes to a real slow crawl as it just overloads our storage system.

There are options to limit the running indexes at the same time etc but that just creates (from the user point of view) weird data as report a then can have data but report b doesn't as yet.
What is the advised way of doing this?

Because we have put in a request for extra temp IO (we are in a private cloud) but the supplier isn't hopeful about this as he thinks that the more data we save the more I/O is needed, and we are planning on retaining about 10 years of data.
Our hope was that we could be online in a week for this worst disaster recovery, but with the current calculations we would have non-logical data for about 3 months.

We are currently thinking we can go 2 ways : 
- Either we create a hot and not so hot database schema, but this would require a major application change.
- Or we would look for a sort of drip-feed mechanism. Where we restore the data in a ravendb without indexes and then drip-feed the data by etl task (does it offer something like this) so we just pump the data day-by-day in reverse into a ravendb with indexes, so we allmost immediately have the latest data back and the older data will follow.

The problem is that for our application to make sense we need an object to be fully indexed over multiple indexes and that is not the way the current indexing works, the current indexing just wants to go as fast as possible through the data.

P.s. keep in mind that this is a worst case disaster recovery scenario, there are about 9 other scenarios in front of this like restoring database backups with indexes, restoring vm's, restoring hdd contents etc. etc. etc.
Basically as long as we have the indexes there is not a real problem.

**Comment by georgiosd** [comment]
I'm still not clear what the disaster plan is here though? Can you elaborate on that? Like the process and/or infrastructure that enables it?

What about using `OutputReduceToCollection` - I'd have to confirm but I think those documents stick around even if the index is lost and I *think* they would also get replicated.

**Comment by Gomez12** [comment]
This is basically the worst and last recovery plan.

Think Azure is completely down (where we keep our regular backups)
Think AWS is completely down (where we keep our long-term backups)
Think our cloud-provider is completely down (so there are no vm backups)
etc. etc. etc.

All we have left is a last resort smuggler export which we do once a week to a local drive.
So we rent an infrastructure on Baidu (as that is the only cloud operator still running) and we restore the smuggler export there.

It is mostly a theoretical scenario, but we are required to have a plan for it by regulations, and the problem also comes to the top if at one time or another we have to reindex the complete data.

The problem is basically that we drip-feed the data in regular business and that we have a more than capable infrastructure for that. 
But we will basically overload any IO system (for a reasonable price) with a complete reindex of all indexes of 10 years of drip-feeding.
And then eventual consistency will bite you in the ass as the data is (basically) a mess until the indexing is completely done, 

Eventual consistency is perfect if it is for seconds, or hours, or even days. But when it starts to go to weeks then you will have problems.

OutputReduceToCollection are only some indexes, a lot of them are simple map-indexes which in a json-Backup are not included (only the definitions)

**Comment by georgiosd** [comment]
You can still apply the same technique as OutputReduceToCollection for non
map-reduce indexes to keep a document level copy but then why keep a
smuggler export and not a full db backup?

Also, have you considered a failover cluster?

By the way, there are tons of other cloud operators than the big 3 - we
work with Scaleway a lot.

**Comment by krivulcik** [comment, marked as answer]
If you can't tolerate reindexing time, you probably need snapshots. These include indexed data, which is restored as-is.

https://ravendb.net/docs/article-page/5.4/csharp/client-api/operations/maintenance/backup/backup#snapshot

It's an Enterprise-level feature, but it might be worth the price for you (if you don't already pay for an Enterprise license). The other solutions seem to be even more expensive.

An alternative approach would be to restore the database on a powerful on-premise server and then move the entire database to the cloud. We have found that bare-metal hardware is many times faster than cloud. In similar scenarios, it's maybe worth exploring - in our case, the import/indexing time is cut by a factor of 2-5 compared to our optimized cloud environment and possibly by an order of magnitude compared to the ordinary cloud environment.

**Reply by georgiosd** [reply]
Not to mention that bare metal is actually much cheaper. Scaleway has the "elastic metal" offer that you can pay per hour and spin up quickly.

**Comment by Gomez12** [comment]
Ok let me rephrase it then as you seem to concentrate on another problem than which i am trying to solve.

It is a problem of scaling. If i have a system which i drip-feed daily, then i will scale it to easily handle a reindexing of a week of data, and maybe uncomfortably but workable a month of data.

But if I need a complete reindex of 10 years of data then a factor of 2-5 faster will help me some as it reduces my (10 years with 12 month a year =) 120x uncomfortable load to about 24x uncomfortable load.
But it requires basically the highest up-to-date bare-metal servers which need to be replaced and which need maintenance etc. etc.
It is worth considering.

But I would think it requires more a different strategy. If we're thinking disaster recovery than we have a different than usual scenario, we have high priority on that the latest / current data is shown and we have constant lower priority as we go further back in time.
It is eventual consistency, but it is eventual consistency with a certain ordering.

Basically it is something like this : 
- We would first need the customers
- Then the products
- Then the orders
- Then the orderlines
Basically I can say on reindexing that customers has high prio, products has high prio.
But within the orders and orderlines there is also an inherent order in the case of disaster recovery.

If I can have my customers and products reindexed and for example the last 2 days for orders and orderlines then I have no problem with about 90% of customers, if i can then backfill it to a month in a days time, then I have no problem with 95% of my customers.
Basically only 0,0000001% of my customers looks at the full 10 years of data. In day-to-day operations it is essentially free to keep this data. But in a crash then you have priority.

@georgiosd the problem isn't the hardware. The problem is that I can create data over time which I can't restore in a reasonable time-period but which can have inherent priority's.
If you are starting to think about petabytes or larger storage than that is no problem over time, but it is a problem in a disaster recovery. Or for example if there comes a bug which requires a total reindexing.

@krivulcik : Snapshots seem a perfect way for my disaster recovery problem. We have an enterprise so that shouldn't be a problem. I will start to investigate if we can do incremental snapshots with a restart every month or so.

**Reply by georgiosd** [reply]
Right, but it doesn't sound like that problem can be solved with how you are trying to solve it. 

From everyone's comments, the possible paths I see are:

- Dont use indexes to calculate this, especially if the data doesn't change after a point. Whether you keep the indexes or not, store the final result as a document.
- Use a failover cluster that gets continuously updated.
- Use snapshots.

Anything I missed? :)

**Reply by Gomez12** [reply]
Imho.
1 is not a reasonable option, a database needs indexes. If you are not using indexes then there are much cheaper and easier options with lots of other features.
2 is basically saying don't plan for disaster, just for small inconveniences. Failover cluster has many upsides, but also a lot of downsides because it is a cluster, user generated data-corruption will for example also update all over your cluster
3 for disaster you will probably need incremental snapshots. Or full snapshots if you have small databases.

**Reply by ayende** [reply]
Your issue is that you are trying to run 10 years of computation in a shorter time frame.
And for map/reduce especially, we have a lot of disk I/O as we update the entries.

There are some modes that may help (running the initial process without transactions, for example), but I would recommend it.

The usual way you'll handle that is to have a full snapshot on a monthly basis, and the daily / hourly incremental on top of that.

Note that if your database is large enough, you may run into problems at an earlier stage. Just _moving_ a hundred GB file from one cloud to another can take a lot of time, for example.
